targets/TARGET_Cypress/TARGET_PSOC6/sleep_api.c

```diff
@@ -26,16 +29,34 @@ void hal_sleep(void)
     Cy_SysPm_Sleep(CY_SYSPM_WAIT_FOR_INTERRUPT);
 }
 
+
+#define DEEP_SLEEP_ENTER_TIMEOUT_US     260000
+
 void hal_deepsleep(void)
 {
 #if DEVICE_LPTICKER
-    if(CY_SYSPM_SUCCESS == Cy_SysPm_DeepSleep(CY_SYSPM_WAIT_FOR_INTERRUPT)) {
-        // Have to make sure PLL clock is restored before continuing.
-        // FLL clock is not used in basic configuration.
-        while(!Cy_SysClk_PllLocked(1)) {
-            // Just wait here.
-        }
+    static volatile uint32_t in_progress = 0;
+
+    // SysPm API is not re-entrant, so make sure not to enter it multiple times.
+    if (core_util_atomic_incr_u32(&in_progress, 1) == 1) {
```

Comment:
`hal_deepsleep()` does not need to be re-entrant. It's called from a critical section (not doing that yields undefined behaviour as it doesn't protect against early Iinterrupts.

Reply:
Actually, HAL requirements are not well documented in Mbed.
I can remove it, should I?

Reply:
I'll go further - you must ;)